### PR TITLE
recreate S3 client if credentials changed

### PR DESCRIPTION
**Pull request by GrigoryPervakov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fix that S3 table holds old credentials after config update.

### CODE REVIEW

File: tests/integration/test_storage_s3/test.py [hunk 1]
```diff
@@ -277,9 +281,9 @@ def test_put_get_with_globs(cluster):
 
 # Test multipart put.
 @pytest.mark.parametrize("maybe_auth,positive", [
-    ("", True),
+    ("", True)
     # ("'minio','minio123',",True), Redirect with credentials not working with nginx.
-    ("'wrongid','wrongkey',", False)
+    # ("'wrongid','wrongkey',", False) ClickHouse crashes in some time after this test, local integration tests run fails.
```
Comment (vdimir) [inline tests/integration/test_storage_s3/test.py:286]: Why this happens? Did it happened before changes?

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -356,6 +340,38 @@ BlockOutputStreamPtr StorageS3::write(const ASTPtr & /*query*/, const StorageMet
         max_single_part_upload_size);
 }
 
+void StorageS3::updateAuthSettings(const Context & context)
+{
+    auto settings = context.getStorageS3Settings().getSettings(uri.uri.toString());
+    if (client && (!access_key_id.empty() || settings == auth_settings))
```
Comment (vdimir) [inline src/Storages/StorageS3.cpp:346]: Can this place be a bottle neck? Comparing settings requires to compare several strings and we perform it in each `StorageS3::read` call.

File: tests/integration/test_storage_s3/test.py (hunk 1 above)
Comment (GrigoryPervakov) [reply tests/integration/test_storage_s3/test.py:286]: Yes, the problem reproduces on the master branch.

File: src/Storages/StorageS3.cpp (hunk 2 above)
Comment (vdimir) [reply src/Storages/StorageS3.cpp:346]: Decided to leave as is, because `StorageS3::read` called not often, once for query, I suppose

**Review by vdimir — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
This is "improvement", not a bugfix.

**Pull request merged**